File: library/cpp/number/next_combination.cpp
#include <bits/stdc++.h>
using namespace std;

/* snippet part start */
int next_combination(int sub){
    int x = sub & -sub, y = sub + x;
    return (((sub & ~y) / x) >> 1) | y;
}

int main() {
    int n=5;
    int k=3;

    int bit = (1<<k)-1;
    for(;bit < (1<<n); bit = next_combination(bit)){
        vector<int> S;
        for(int i=0; i<n; i++){
            if(bit & (1<<i)){
                S.push_back(i);
            }
        }

        cout << bit <<":";
        for(int i=0; i<(int)S.size(); i++){
            cout <<" "<< S[i];
        }
        cout << endl;
    }
}
/* snippet part end */

File: library/rust/src/number/next_combination.rs
// unverified

fn next_combination(sub: i64) -> i64 {
    let x = sub & -sub;
    let y = sub + x;
    (((sub & !y) / x) >> 1) | y
}

#[test]
fn test1() {
    let n = 5;
    let k = 3;

    // 5C3
    let mut bit = (1 << k) - 1;
    while bit < (1 << n) {
        let mut s = vec![];
        for i in 0..n {
            if bit & (1 << i) > 0 {
                s.push(i);
            }
        }
        println!("{}: {:05b}", bit, bit);
        for i in 0..s.len() {
            print!("{} ", s[i]);
        }
        println!("");

        bit = next_combination(bit);
    }
}